sympy/solvers/solveset.py

```diff
@@ -998,6 +1003,148 @@ def _solveset(f, symbol, domain, _check=False):
+
+
+def _invert_modular(modterm, rhs, symbol):
+    """
+    Helper function to invert modular equation.
+    ``Mod(a, m) - rhs = 0``
+
+    Generally it is inverted as (a, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)).
+    More simplified form will be returned if possible.
+
+    Parameters
+    ==========
+
+    modterm, rhs : Expr
+        The modular equation to be inverted, ``modterm - rhs = 0``
+
+    symbol : Symbol
+        The variable in the equation to be inverted.
+
+    Returns
+    =======
+
+    A tuple (f_x, y) is being returned where f_x is modular independent function
+    of symbol and y being set of values f_x can have.
+
+    """
+    a, m = modterm.args
+    n = Dummy('n')
+
+    if a is symbol:
+        return symbol, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)
+
+    if a.is_Add:
+        # g + h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.Zero:
+            return _invert_modular(Mod(h, m), (rhs - Mod(g, m)) % m, symbol)
+
+    if a.is_Mul:
+        # g*h = a
+        g, h = a.as_independent(symbol)
+        if g is not S.One:
+            return _invert_modular(Mod(h, m), (rhs*invert(g, m)) % m, symbol)
+
+    if a.is_Pow:
+        # base**expo = a
+        base, expo = a.args
+        if expo.has(symbol) and not base.has(symbol):
+            return expo, FiniteSet(discrete_log(m, rhs, a.base))
+
+    return a, ImageSet(Lambda(n, m*n + rhs % m), S.Integers)
+
+
+def _solve_modular(f, symbol, domain):
+    r"""
+    Helper function for solving modular equations of type ``A - Mod(B, C) = 0``,
+    where A can or cannot be a function of symbol, B is surely a function of
+    symbol and C is an integer.
+
+    Currently only cases when A is not a function of symbol is been implemented.
+
+    Parameters
+    ==========
+
+    f : Expr
+        The modular equation to be solved, ``f = 0``
+
+    symbol : Symbol
+        The variable in the equation to be solved.
+
+    domain : Set
+        A set over which the equation is solved.
+
+    Returns
+    =======
+
+    A set of solutions satisfying the given modular equation.
+    A ``ConditionSet`` if the equation is unsolvable.
+
+    Examples
+    ========
+
+    >>> from sympy.solvers.solveset import _solve_modular as solve_modulo
+    >>> from sympy import S, Symbol, sin
+    >>> from sympy.core.mod import Mod
+    >>> x = Symbol('x')
+    >>> solve_modulo(Mod(5*x - 8, 7) - 3, x, S.Reals)
+    ImageSet(Lambda(_n, 7*_n + 5), Integers)
+    >>> solve_modulo(Mod(x**2 - 2*x + 1, 7) - 3, x, S.Complexes)
+    ImageSet(Lambda(_n, {1 - sqrt(7*_n + 3), sqrt(7*_n + 3) + 1}), Integers)
+    >>> solve_modulo(Mod(sin(x), 7) - 3, x, S.Reals) # not solvable specifically
+    ImageSet(Lambda(_n, ConditionSet(x, Eq(-7*_n + sin(x) - 3, 0), Reals)), Integers)
+
+    """
+    # extract modterm and g_ys from f
+    modterm = list(f.atoms(Mod))[0]
+    t = Dummy('t', integer=True)
+    f = f.xreplace({modterm: t})
+    g_ys = _solveset(f, t, S.Integers)
```

Comment:
Consider an equation like this where `t` is representing modterm.
```
t**2 - 1 = 0
```
then clearly t = +/-1
thats what g_ys is meant to be to get the values t can have. This shouldn't be incorporated in `_invert_modular` because it is preprocessing and `_invert_modular` is not meant to invert equation like `t**2 - 1`.

Reply:
Okay, then you might want to add a few tests for these types.

Reply:
Added tests related to this also.